optimize/protocol.py

```diff
@@ -0,0 +1,254 @@
+
+    def u_interp(u_n: Array, lam: Float) -> Float:
+        return np.nan_to_num(np.interp(lam, lambdas, u_n), nan=+np.inf, posinf=+np.inf)
+
+    @jit
+    def vec_u_interp(lam: Float) -> Array:
+        return vmap(u_interp, (1, None))(u_kn, lam)
+
+    @jit
+    def vec_delta_u(from_lam: Float, to_lam: Float) -> Array:
+        """+inf minus +inf -> 0, rather than +inf minus +inf -> nan"""
+        raw_delta_u = vec_u_interp(to_lam) - vec_u_interp(from_lam)
+        return np.nan_to_num(raw_delta_u)
+
+    return u_interp, vec_u_interp, vec_delta_u
+
+
+def construct_work_stddev_estimator(
+    reference_log_weights_n: Array, vec_u: Callable, vec_delta_u: Callable
+) -> WorkStddevEstimator:
+    """Construct reweighted estimator for stddev from a collection of reference samples"""
+
+    def work_stddev_estimator(prev_lam: Float, next_lam: Float) -> Float:
+        target_logpdf_n = -vec_u(prev_lam)
+        delta_us = vec_delta_u(prev_lam, next_lam)
+
+        stddev_estimate = reweighted_stddev(
+            f_n=delta_us,
+            target_logpdf_n=target_logpdf_n,
+            source_logpdf_n=reference_log_weights_n,
+        )
+
+        return stddev_estimate
+
+    return work_stddev_estimator
+
+
+def reweighted_stddev(f_n: Array, target_logpdf_n: Array, source_logpdf_n: Array) -> Float:
+    """Compute reweighted estimate of
+    stddev(f(x)) under x ~ p_target
+    based on samples   x ~ p_source
+
+    where
+        p_target(x) = exp(target_logpdf(x)) / Z_target
+
+    using samples from a different source
+        x_n ~ p_source
+        where
+        p_source(x) = exp(source_logpdf(x)) / Z_source
+
+    The inputs are arrays "{fxn_name}_n" containing the result of
+    calling each fxn on a fixed array of samples:
+
+    * f_n = [f(x_n) for x_n in samples]
+    * target_logpdf_n = [target_logpdf(x_n) for x_n in samples]
+    * source_logpdf_n = [source_logpdf(x_n) for x_n in samples]
+    """
+
+    log_weights_n = target_logpdf_n - source_logpdf_n
+    weights = np.exp(log_weights_n - logsumexp(log_weights_n)).flatten()
+
+    f_mean = np.sum(weights * f_n)
+    squared_deviations = (f_n - f_mean) ** 2
+
+    # sanitize 0 * inf -> 0 (instead of nan)
+    weighted_squared_deviations = weights * squared_deviations
+    sanitized = np.nan_to_num(weighted_squared_deviations, nan=0)
+    stddev = np.sqrt(np.sum(sanitized))
+
+    return stddev
+
+
+def construct_heuristic_lambda_pair_assessor(work_stddev_estimator) -> StepAssessor:
+    """Construct a function f(prev_lam, trial_next_lam) where bisection search on second argument
+    can be used to select next_lam so that p(x|next_lam) is a specified "distance" from p(x|prev_lam)"""
+
+    def assess_lambda_pair(prev_lam, next_lam, desired_stddev=1.0, max_step=0.25):
+        """if (next_lam - prev_lam <= max_step), compute (max(forward_stddev, reverse_stddev) - desired_stddev)"""
+        too_far = next_lam - prev_lam > max_step
+        if too_far:
+            return +np.inf
```

Comment:
Arbitrary speed limit imposed by default, to explicitly prevent any big lambda increments.

If applied to solvent decoupling stage, would bias the optimization result to include more intermediate states between lam=0.6 and lam=1.0 than it would otherwise include.

Should we retain this speed limit? Modify its value?

In current manual protocol, a step of size 0.4 is intentional https://github.com/proteneer/timemachine/blob/41c5a0cdbd38f8cec27b7f1bb973a35a83cc0c58/fe/free_energy_rabfe.py#L252 and in the new pre-optimized protocol, this speed-limit was omitted / a step of size ~0.4 was obtained https://github.com/proteneer/timemachine/blob/41c5a0cdbd38f8cec27b7f1bb973a35a83cc0c58/fe/free_energy_rabfe.py#L295